### ISSUE TITLE
Help Needed on Out of Memory Issue

### ISSUE THREAD

**Issue by vt911001** [microsoft/tensorflow-directml#294]
Hi,

I have tried to run SSD MobilenetV1 model from [http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_2018_01_28.tar.gz](http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_2018_01_28.tar.gz) with **tensorflow-directml** precompiled C libraries available in [https://github.com/microsoft/tensorflow-directml/releases/tag/v1.15.5.dev210429](https://github.com/microsoft/tensorflow-directml/releases/tag/v1.15.5.dev210429).

However, I am getting OOM issue when running the model. The snippet of the output is as below:
```
2021-06-28 09:23:32.032871: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2021-06-28 09:23:32.035434: I tensorflow/stream_executor/platform/default/dso_loader.cc:99] Successfully opened dynamic library C:\Users\vince\Downloads\Release/directml.adbd007a01a52364381a1c71ebb6fa1b2389c88d.dll
2021-06-28 09:23:32.166394: I tensorflow/core/common_runtime/dml/dml_device_cache.cc:249] DirectML device enumeration: found 1 compatible adapters.
2021-06-28 09:23:32.232103: I tensorflow/core/common_runtime/dml/dml_device_cache.cc:185] DirectML: creating device on adapter 0 (NVIDIA GeForce GTX 950)
2021-06-28 09:23:32.299033: I tensorflow/stream_executor/platform/default/dso_loader.cc:99] Successfully opened dynamic library Kernel32.dll
...
2021-06-28 09:24:26.743588: I tensorflow/core/common_runtime/bfc_allocator.cc:935]      Summary of in-use Chunks by size: 
2021-06-28 09:24:26.743597: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 147330 Chunks of size 256 totalling 35.97MiB
2021-06-28 09:24:26.743604: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 704 Chunks of size 512 totalling 352.0KiB
2021-06-28 09:24:26.743609: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 69 Chunks of size 768 totalling 51.8KiB
2021-06-28 09:24:26.743615: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 9 Chunks of size 1024 totalling 9.0KiB
2021-06-28 09:24:26.743620: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 3 Chunks of size 1280 totalling 3.8KiB
2021-06-28 09:24:26.743626: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 19 Chunks of size 2048 totalling 38.0KiB
2021-06-28 09:24:26.743632: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 6 Chunks of size 2304 totalling 13.5KiB
2021-06-28 09:24:26.743637: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 3584 totalling 3.5KiB
2021-06-28 09:24:26.743643: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 4 Chunks of size 4096 totalling 16.0KiB
2021-06-28 09:24:26.743648: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 2 Chunks of size 4608 totalling 9.0KiB
2021-06-28 09:24:26.743654: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 4864 totalling 4.8KiB
2021-06-28 09:24:26.743660: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 63201 Chunks of size 7680 totalling 462.90MiB
2021-06-28 09:24:26.743665: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 101 Chunks of size 7936 totalling 782.8KiB
2021-06-28 09:24:26.743671: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 69 Chunks of size 8192 totalling 552.0KiB
2021-06-28 09:24:26.743676: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 39 Chunks of size 8448 totalling 321.8KiB
2021-06-28 09:24:26.743682: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 51 Chunks of size 8704 totalling 433.5KiB
2021-06-28 09:24:26.743687: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 62 Chunks of size 8960 totalling 542.5KiB
2021-06-28 09:24:26.743693: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 117 Chunks of size 9216 totalling 1.03MiB
2021-06-28 09:24:26.743698: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 189 Chunks of size 9472 totalling 1.71MiB
2021-06-28 09:24:26.743704: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 2960 Chunks of size 9728 totalling 27.46MiB
2021-06-28 09:24:26.743709: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 54 Chunks of size 9984 totalling 526.5KiB
2021-06-28 09:24:26.743715: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 38 Chunks of size 10240 totalling 380.0KiB
2021-06-28 09:24:26.743720: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 51 Chunks of size 10496 totalling 522.8KiB
2021-06-28 09:24:26.743726: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 94 Chunks of size 10752 totalling 987.0KiB
2021-06-28 09:24:26.743732: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 122 Chunks of size 11008 totalling 1.28MiB
2021-06-28 09:24:26.743737: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 290 Chunks of size 11264 totalling 3.12MiB
2021-06-28 09:24:26.743743: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 49 Chunks of size 11520 totalling 551.3KiB
2021-06-28 09:24:26.743748: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 616 Chunks of size 11776 totalling 6.92MiB
2021-06-28 09:24:26.743754: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 27 Chunks of size 12032 totalling 317.3KiB
2021-06-28 09:24:26.743759: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 55 Chunks of size 12288 totalling 660.0KiB
2021-06-28 09:24:26.743766: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 72 Chunks of size 12544 totalling 882.0KiB
2021-06-28 09:24:26.743772: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 174 Chunks of size 12800 totalling 2.12MiB
2021-06-28 09:24:26.743778: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 494 Chunks of size 13056 totalling 6.15MiB
2021-06-28 09:24:26.743783: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 943 Chunks of size 13312 totalling 11.97MiB
2021-06-28 09:24:26.743789: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 22 Chunks of size 13568 totalling 291.5KiB
2021-06-28 09:24:26.743794: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 129 Chunks of size 13824 totalling 1.70MiB
2021-06-28 09:24:26.743800: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 17 Chunks of size 14080 totalling 233.8KiB
2021-06-28 09:24:26.743805: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 44 Chunks of size 14336 totalling 616.0KiB
2021-06-28 09:24:26.743810: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 132 Chunks of size 14592 totalling 1.84MiB
2021-06-28 09:24:26.743816: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 304 Chunks of size 14848 totalling 4.30MiB
2021-06-28 09:24:26.743821: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1399 Chunks of size 15104 totalling 20.15MiB
2021-06-28 09:24:26.743827: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 69491 Chunks of size 15360 totalling 1017.93MiB
2021-06-28 09:24:26.743833: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 32 Chunks of size 15616 totalling 488.0KiB
2021-06-28 09:24:26.743839: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 45 Chunks of size 15872 totalling 697.5KiB
2021-06-28 09:24:26.743844: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 16 Chunks of size 16128 totalling 252.0KiB
2021-06-28 09:24:26.743849: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 25 Chunks of size 16384 totalling 400.0KiB
2021-06-28 09:24:26.743855: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 14 Chunks of size 16640 totalling 227.5KiB
2021-06-28 09:24:26.743860: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 14 Chunks of size 16896 totalling 231.0KiB
2021-06-28 09:24:26.743865: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 19 Chunks of size 17152 totalling 318.3KiB
2021-06-28 09:24:26.743871: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 462 Chunks of size 17408 totalling 7.67MiB
2021-06-28 09:24:26.743876: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 12 Chunks of size 17664 totalling 207.0KiB
2021-06-28 09:24:26.743882: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 20 Chunks of size 17920 totalling 350.0KiB
2021-06-28 09:24:26.743887: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 14 Chunks of size 18176 totalling 248.5KiB
2021-06-28 09:24:26.743892: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 21 Chunks of size 18432 totalling 378.0KiB
2021-06-28 09:24:26.743898: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 25 Chunks of size 18688 totalling 456.3KiB
2021-06-28 09:24:26.743904: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 14 Chunks of size 18944 totalling 259.0KiB
2021-06-28 09:24:26.743909: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 20 Chunks of size 19200 totalling 375.0KiB
2021-06-28 09:24:26.743915: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 214 Chunks of size 19456 totalling 3.97MiB
2021-06-28 09:24:26.743920: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 11 Chunks of size 19712 totalling 211.8KiB
2021-06-28 09:24:26.743925: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 17 Chunks of size 19968 totalling 331.5KiB
2021-06-28 09:24:26.743931: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 9 Chunks of size 20224 totalling 177.8KiB
2021-06-28 09:24:26.743936: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 24 Chunks of size 20480 totalling 480.0KiB
2021-06-28 09:24:26.743942: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 32 Chunks of size 20736 totalling 648.0KiB
2021-06-28 09:24:26.743949: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 54 Chunks of size 20992 totalling 1.08MiB
2021-06-28 09:24:26.743954: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 13 Chunks of size 21248 totalling 269.8KiB
2021-06-28 09:24:26.743959: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 64 Chunks of size 21504 totalling 1.31MiB
2021-06-28 09:24:26.743965: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 8 Chunks of size 21760 totalling 170.0KiB
2021-06-28 09:24:26.743970: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 19 Chunks of size 22016 totalling 408.5KiB
2021-06-28 09:24:26.743979: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 15 Chunks of size 22272 totalling 326.3KiB
2021-06-28 09:24:26.743985: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 25 Chunks of size 22528 totalling 550.0KiB
2021-06-28 09:24:26.743990: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 62 Chunks of size 22784 totalling 1.35MiB
2021-06-28 09:24:26.743996: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 511 Chunks of size 23040 totalling 11.23MiB
2021-06-28 09:24:26.744001: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 11 Chunks of size 23296 totalling 250.3KiB
2021-06-28 09:24:26.744007: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 22 Chunks of size 23552 totalling 506.0KiB
2021-06-28 09:24:26.744012: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 4 Chunks of size 23808 totalling 93.0KiB
2021-06-28 09:24:26.744018: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 15 Chunks of size 24064 totalling 352.5KiB
2021-06-28 09:24:26.744023: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 3 Chunks of size 24320 totalling 71.3KiB
2021-06-28 09:24:26.744029: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 11 Chunks of size 24576 totalling 264.0KiB
2021-06-28 09:24:26.744034: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 24 Chunks of size 24832 totalling 582.0KiB
2021-06-28 09:24:26.744040: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 113 Chunks of size 25088 totalling 2.70MiB
2021-06-28 09:24:26.744046: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 7 Chunks of size 25344 totalling 173.3KiB
2021-06-28 09:24:26.744051: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 19 Chunks of size 25600 totalling 475.0KiB
2021-06-28 09:24:26.744056: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 9 Chunks of size 25856 totalling 227.3KiB
2021-06-28 09:24:26.744062: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 11 Chunks of size 26112 totalling 280.5KiB
2021-06-28 09:24:26.744068: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 11 Chunks of size 26368 totalling 283.3KiB
2021-06-28 09:24:26.744073: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 14 Chunks of size 26624 totalling 364.0KiB
2021-06-28 09:24:26.744078: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 16 Chunks of size 26880 totalling 420.0KiB
2021-06-28 09:24:26.744084: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 79 Chunks of size 27136 totalling 2.04MiB
2021-06-28 09:24:26.744089: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 6 Chunks of size 27392 totalling 160.5KiB
2021-06-28 09:24:26.744095: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 13 Chunks of size 27648 totalling 351.0KiB
2021-06-28 09:24:26.744100: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 8 Chunks of size 27904 totalling 218.0KiB
2021-06-28 09:24:26.744106: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 13 Chunks of size 28160 totalling 357.5KiB
2021-06-28 09:24:26.744111: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 11 Chunks of size 28416 totalling 305.3KiB
2021-06-28 09:24:26.744116: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 26 Chunks of size 28672 totalling 728.0KiB
2021-06-28 09:24:26.744122: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 10 Chunks of size 28928 totalling 282.5KiB
2021-06-28 09:24:26.744128: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 51 Chunks of size 29184 totalling 1.42MiB
2021-06-28 09:24:26.744134: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 15 Chunks of size 29440 totalling 431.3KiB
2021-06-28 09:24:26.744140: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 23 Chunks of size 29696 totalling 667.0KiB
2021-06-28 09:24:26.744145: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 11 Chunks of size 29952 totalling 321.8KiB
2021-06-28 09:24:26.744150: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 20 Chunks of size 30208 totalling 590.0KiB
2021-06-28 09:24:26.744156: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 21 Chunks of size 30464 totalling 624.8KiB
2021-06-28 09:24:26.744161: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 30720 totalling 30.0KiB
2021-06-28 09:24:26.744167: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 32768 totalling 32.0KiB
2021-06-28 09:24:26.744172: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 36864 totalling 36.0KiB
2021-06-28 09:24:26.744177: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 49152 totalling 48.0KiB
2021-06-28 09:24:26.744183: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 2 Chunks of size 65536 totalling 128.0KiB
2021-06-28 09:24:26.744188: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 98304 totalling 96.0KiB
2021-06-28 09:24:26.744194: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 2 Chunks of size 131072 totalling 256.0KiB
2021-06-28 09:24:26.744199: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 2 Chunks of size 262144 totalling 512.0KiB
2021-06-28 09:24:26.744205: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 279552 totalling 273.0KiB
2021-06-28 09:24:26.744210: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 294912 totalling 288.0KiB
2021-06-28 09:24:26.744216: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 524288 totalling 512.0KiB
2021-06-28 09:24:26.744221: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 3 Chunks of size 559104 totalling 1.60MiB
2021-06-28 09:24:26.744227: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 6 Chunks of size 1048576 totalling 6.00MiB
2021-06-28 09:24:26.744232: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 1118208 totalling 1.07MiB
2021-06-28 09:24:26.744238: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 2 Chunks of size 1179648 totalling 2.25MiB
2021-06-28 09:24:26.744244: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 2097152 totalling 2.00MiB
2021-06-28 09:24:26.744249: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 2236416 totalling 2.13MiB
2021-06-28 09:24:26.744254: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 4194304 totalling 4.00MiB
2021-06-28 09:24:26.744260: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 4718592 totalling 4.50MiB
2021-06-28 09:24:26.744265: I tensorflow/core/common_runtime/bfc_allocator.cc:938] 1 Chunks of size 10957568 totalling 10.45MiB
2021-06-28 09:24:26.744270: I tensorflow/core/common_runtime/bfc_allocator.cc:942] Sum Total of in-use chunks: 1.66GiB
2021-06-28 09:24:26.744277: I tensorflow/core/common_runtime/bfc_allocator.cc:944] total_region_allocated_bytes_: 1789375232 memory_limit_: 1789375284 available bytes: 52 curr_region_allocation_bytes_: 3578750976
2021-06-28 09:24:26.744285: I tensorflow/core/common_runtime/bfc_allocator.cc:950] Stats: 
Limit:                  1789375284
InUse:                  1784465664
MaxInUse:               1788396032
NumAllocs:                  821973
MaxAllocSize:             11416576

2021-06-28 09:24:26.758302: W tensorflow/core/common_runtime/bfc_allocator.cc:445] ****************************************************************************************************
2021-06-28 09:24:26.758346: W tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at dml_kernel_context.cc:167 : Resource exhausted: OOM when allocating tensor with shape[1,64,150,150] and type float on /job:localhost/replica:0/task:0/device:DML:0 by allocator DmlAllocator
2 root error(s) found.
  (0) Resource exhausted: OOM when allocating tensor with shape[1,64,150,150] and type float on /job:localhost/replica:0/task:0/device:DML:0 by allocator DmlAllocator
	 [[{{node FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_1_pointwise/BatchNorm/batchnorm/add_1}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

	 [[Postprocessor/BatchMultiClassNonMaxSuppression/map/while/LoopCond/_1539]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

  (1) Resource exhausted: OOM when allocating tensor with shape[1,64,150,150] and type float on /job:localhost/replica:0/task:0/device:DML:0 by allocator DmlAllocator
	 [[{{node FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_1_pointwise/BatchNorm/batchnorm/add_1}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

0 successful operations.
0 derived errors ignored.
```

# System Information

Host System
--------------------------------------------------------------------------------
Windows 10 Version  : Windows 10 Pro 64-bit (10.0, Build 19041) (19041.vb_release.191206-1406)
Processor           : AMD Ryzen 5 3600 6-Core Processor               (12 CPUs), ~3.6GHz
Memory              : 16384MB RAM
DirectX Version     : DirectX 12

Python Environment
--------------------------------------------------------------------------------
Python Version      : 3.6.8
TensorFlow-DirectML : 1.15.5.dev210429

DirectX Device
--------------------------------------------------------------------------------
Description         : NVIDIA GeForce GTX 950
Manufacturer        : NVIDIA
Chip Type           : GeForce GTX 950
Dedicated Memory    : 2007 MB
Driver Version      : 27.21.14.5167
Driver Model        : WDDM 2.7
Driver Date         : 5/7/2020 8:00:00 AM
Feature Levels      : 12_1,12_0,11_1,11_0,10_1,10_0,9_3,9_2,9_1

# Repro Details

**Describe the current behavior**
OOM after running the sample for some time(>10s)

**Describe the expected behavior**
No OOM issue

**Code to reproduce the issue**
The attached VS2019 projects can be compiled with std c++17, opencv and tensorflow-directml C libraries.
[https://drive.google.com/drive/folders/126U92nV160TaWZLOUSH5FYUK6-mpXCZK?usp=sharing](https://drive.google.com/drive/folders/126U92nV160TaWZLOUSH5FYUK6-mpXCZK?usp=sharing)

The executable can be run as below:
`tf_directml_ssdmobilenetv1.exe [video file]`

FYI, I have used [https://github.com/serizba/cppflow/tree/243ff2fc4e33632b91676cad7d6cfc3c92308601](https://github.com/serizba/cppflow/tree/243ff2fc4e33632b91676cad7d6cfc3c92308601) in the sample codes.

**Other info / logs**
The complete logs along with system info are located in [https://drive.google.com/drive/folders/126U92nV160TaWZLOUSH5FYUK6-mpXCZK?usp=sharing](https://drive.google.com/drive/folders/126U92nV160TaWZLOUSH5FYUK6-mpXCZK?usp=sharing) as well.

Besides, I have tested the exact same codes but by linking it with normal **tensorflow-cpu** C libraries and I dont see the OOM issue.

Thanks,
Vincent

**Comment by jstoecker** [comment]
Hi Vincent,

Unfortunately there's not much you can do except run models with a smaller batch size if you encounter OOM issues. I believe this script you're using is for inference, though, so that doesn't leave many options. You have 16GB of system memory available when running with a CPU, but you only have ~1.7GB of dedicated video memory when using your GTX 950. The DirectML backend of TensorFlow requires most tensors to be stored in video memory for efficient access, as using system memory would be prohibitively slow (this is true for CUDA as well).

We did have a separate issue with this model that won't be fixed until the next release as well. We'll take a look to see if the peak memory footprint is unexpected, but I do recall seeing OOM issues with CUDA in these TF object detection models.

**Comment by vt911001** [comment]
Hi Justin,

I am so sorry for the late reply.
BTW, I have tried the exact codes/executable on RTX 2080 with 8GB of memory and I still bumped into similar error.
```
2021-07-28 16:48:05.967788: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2021-07-28 16:48:05.970859: I tensorflow/stream_executor/platform/default/dso_loader.cc:99] Successfully opened dynamic library C:\Users\rbhesve\Downloads\Release/directml.adbd007a01a52364381a1c71ebb6fa1b2389c88d.dll
2021-07-28 16:48:06.106624: I tensorflow/core/common_runtime/dml/dml_device_cache.cc:249] DirectML device enumeration: found 1 compatible adapters.
2021-07-28 16:48:06.180587: I tensorflow/core/common_runtime/dml/dml_device_cache.cc:185] DirectML: creating device on adapter 0 (NVIDIA GeForce RTX 2080)
2021-07-28 16:48:06.259175: I tensorflow/stream_executor/platform/default/dso_loader.cc:99] Successfully opened dynamic library Kernel32.dll
2021-07-28 16:50:31.238919: W tensorflow/core/common_runtime/bfc_allocator.cc:440] Allocator (DmlAllocator) ran out of memory trying to allocate 5.49MiB (rounded to 5760000).  
...
...
...
2021-07-28 16:50:38.791498: I tensorflow/core/common_runtime/bfc_allocator.cc:950] Stats: 
Limit:                  7593787392
InUse:                  7589175808
MaxInUse:               7593157376
NumAllocs:                 3570403
MaxAllocSize:             11279616

2021-07-28 16:50:38.862533: W tensorflow/core/common_runtime/bfc_allocator.cc:445] ****************************************************************************************************
2021-07-28 16:50:38.862564: W tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at dml_kernel_context.cc:167 : Resource exhausted: OOM when allocating tensor with shape[1,64,150,150] and type float on /job:localhost/replica:0/task:0/device:DML:0 by allocator DmlAllocator
2 root error(s) found.
  (0) Resource exhausted: OOM when allocating tensor with shape[1,64,150,150] and type float on /job:localhost/replica:0/task:0/device:DML:0 by allocator DmlAllocator
	 [[{{node FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_1_pointwise/BatchNorm/batchnorm/add_1}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

	 [[Postprocessor/BatchMultiClassNonMaxSuppression/map/while/LoopCond/_1539]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

  (1) Resource exhausted: OOM when allocating tensor with shape[1,64,150,150] and type float on /job:localhost/replica:0/task:0/device:DML:0 by allocator DmlAllocator
	 [[{{node FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_1_pointwise/BatchNorm/batchnorm/add_1}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

0 successful operations.
0 derived errors ignored.
```

The full error log is available in [https://drive.google.com/file/d/1pO5Qr7RK3rwYGxgeav8ejBfC6IIi55sj/view?usp=sharing](https://drive.google.com/file/d/1pO5Qr7RK3rwYGxgeav8ejBfC6IIi55sj/view?usp=sharing).


Thanks,
Vincent

**Comment by jstoecker** [comment]
We made some changes to memory allocation that should improve utilization of dedicated GPU memory, but these changes aren't in pypi.org releases yet.

One thing you can experiment with for now (this won't be necessary in the next release) is setting some environment variables that control how memory is allocated to avoid fragmentation issues. The first variable below sets the maximum single allocation size to 1GB (value in quotes is bytes), which is _probably_ large enough for most models. If the model requires a tensor that's larger than 1GB this won't work and you'll need a large size. The second variable avoids allocating massive amounts of GPU memory upfront. I'm using powershell syntax to set these, so adjust as appropriate if you're using command prompt or want to set them in a python script:

```powershell
$env:TF_DIRECTML_MAX_ALLOC_SIZE="1073741824"
$env:TF_FORCE_GPU_ALLOW_GROWTH="true"
```

I'll try to run this model with our latest changes to see if this is likely to be resolved by the allocator improvements.

**Comment by jstoecker** [comment]
I was able to build and run your C++ sample with OpenCV 4.5.3 and the [TFDML C library from April](https://github.com/microsoft/tensorflow-directml/releases/tag/v1.15.5.dev210429). It runs for me on an RTX 2070 SUPER (also 8GB) _without_ any additional changes, but it definitely doesn't hurt to also set the `TF_FORCE_GPU_ALLOW_GROWTH` variable in the VS environment settings:

![image](https://user-images.githubusercontent.com/3835701/127908334-0c1ca141-5438-4155-a9d2-64b1cb8a0780.png)

I think you're running into video memory fragmentation issues that will be addressed in the next release without having to change any variables.

**Comment by vt911001** [comment]
Hi Justin,

Let me try and let you know.
Thanks alot for your immense help.


Thanks,
Vincent

**Comment by vt911001** [comment]
Hi Justin,

I have set the following ENV in the cmd prior to running the sample app.
```
set TF_DIRECTML_MAX_ALLOC_SIZE=1073741824
set TF_FORCE_GPU_ALLOW_GROWTH=true
```
Unfortunately, I still bumped into the OOM errors after running the app for few minutes.

![RTX2080_loaded_with_env](https://user-images.githubusercontent.com/68520732/128138538-0ba8655c-8772-4e1f-8510-9bc8071e326e.PNG)
![RTX2080_failed_with_env](https://user-images.githubusercontent.com/68520732/128138569-b4d1455c-661d-4375-b904-a866af895329.PNG)


Thanks,
Vincent

**Comment by jstoecker** [comment]
Hi Vincent,

Thanks for testing that. Going OOM only after a sustained period of time sounds a lot like a leak of some sort. We'll take another and get back to you.

**Comment by jstoecker** [comment]
Hi Vincent,

This will be fixed in the next release. I greatly appreciate you reporting this, and the repro app you provided was a very helpful!

Thanks,
Justin

**Comment by vt911001** [comment]
Hi Justin,

No problem, thanks alot for your time for this.
Looking forward to next release.


Thanks,
Vincent
